### DISCUSSION TITLE
File handlers could receive window number

### DISCUSSION THREAD

**Discussion by vieiro** [webui-dev/webui#486] [Ideas]
When you're programming in C (or Rust, etc.) it's easy to create a C function with the following signature...
```C
const void*(*handler)(const char* filename, int* length);
```
... and set it as the file handler of a given window with `webui_set_file_handler(window, handler)`. In C it's easy to create different functions for different windows.

But when you use another (dynamic) language and you use a compiled version of webui then it may be difficult to create a new C function for each window.  In these programming languages you probably want to create a _single file handler_ and compile it as a library, and then let the file handler act differently depending on which window it's handling. Sadly the file-handlers have no idea which window they're serving.

An enhancement could be to modify the file handler signature to:
```C
const void*(*handler)(size_t window, const char* filename, int* length);
```
this is, send the window `num` to the file handler as a first argument.

The change will require modifying the API (or augmenting it with a different file handler function, say `webui_set_file_handler_window`), but should be easy to implement otherwise.

I can submit a PR if this you consider this useful.

**Comment by AlbertShown** [comment]
Yes, this is very useful, you can create a PR 👍
But it should be `webui_set_file_handler_window()` as you suggested because modifying the existing `webui_set_file_handler()` will break examples and all wrappers.

Call point:
https://github.com/webui-dev/webui/blob/ead99700e99d25500f8a4a4b5ca9efcc6f28b47d/src/webui.c#L4198

**Reply by vieiro** [reply]
I'll add a PR during this week.

Just one question, if I call `webui_set_file_handler` and later on `webui_set_file_handler_window` I assume the last one takes precedence, right? 

What I mean is that:
- calling `webui_set_file_handler_window(window, handler)` implies a call to `webui_set_file_handler(window, NULL)` , resetting any previous file handler.
- and calling `webui_set_file_handler(window, handler)` implies a call `webui_set_file_handler_window(window, NULL)`, resetting any previous "window" file handler.

Is this the required behaviour?

**Reply by AlbertShown** [reply]
Yes, that's right, because the HTTP server should only return one result, so one handler should be used only as you suggested.

**Comment by vieiro** [comment]
Here's a proposed implementation: https://github.com/webui-dev/webui/pull/488
